File: src/dataset/javascript/bowling_exercism-ThomasZumsteg/gemini/GeminiFlash_bowling_v2.js
class Bowling {
    constructor() {
        this.rolls = [];
        this.frame = 0;
        this.extra = 0;
        this.rollCount = 0;
    }

    gameOver() {
        return this.frame >= 10 && this.extra <= 0;
    }

    roll(pins) {
        if (this.gameOver()) {
            throw new Error('Cannot roll after game is over');
        }

        if (pins < 0) {
            throw new Error('Negative roll is invalid');
        }

        const lastRoll = this.rolls.length > 0 ? this.rolls[this.rolls.length - 1] : 0;
        if (pins > 10 || (this.rollCount % 2 === 1 && lastRoll + pins > 10)) {
            throw new Error('Pin count exceeds pins on the lane');
        }

        if (this.frame === 9) {
            if (this.rollCount % 2 === 0 && pins === 10) {
                this.extra = 2;
            } else if (this.rollCount % 2 === 1 && lastRoll + pins === 10) {
                this.extra = 1;
            }
        } else if (this.frame > 9) {
            this.extra--;
        }

        this.rolls.push(pins);
        this.rollCount++;

        if (pins === 10) {
            this.frame++;
        } else if (this.rollCount % 2 === 0) {
            this.frame++;
        }
    }

    score() {
        if (!this.gameOver()) {
            throw new Error('Score cannot be taken until the end of the game');
        }

        let result = 0;
        let i = 0;
        for (let frame = 0; frame < 10; frame++) {
            const firstRoll = this.rolls[i];
            const secondRoll = this.rolls[i + 1];

            result += firstRoll + secondRoll;

            if (firstRoll === 10 || firstRoll + secondRoll === 10) {
                result += this.rolls[i + 2];
            }

            i += (firstRoll === 10 ? 1 : 2);
        }
        return result;
    }
}

module.exports = Bowling;

File: src/dataset/python/bowling_exercism-ThomasZumsteg/gemini/GeminiFlash_bowling_v2.py
class BowlingGame(object):
    def __init__(self):
        self._rolls = []
        self._frame = 0
        self._first = True
        self._extra_rolls = 0

    def roll(self, pins):
        if self._game_over():
            raise IndexError("Game Over")
        if not (0 <= pins <= 10):
            raise ValueError("Cannot score {}".format(pins))
        if not self._first and pins + self._rolls[-1] > 10:
            raise ValueError("Cannot score {} and {}".format(self._rolls[-1], pins))

        self._rolls.append(pins)
        self._update_state(pins)

    def score(self):
        if not self._game_over():
            raise IndexError("Game in progress")

        total = 0
        frame_index = 0
        for frame in range(10):
            if self._is_strike(frame_index):
                total += 10 + self._strike_bonus(frame_index)
                frame_index += 1
            elif self._is_spare(frame_index):
                total += 10 + self._spare_bonus(frame_index)
                frame_index += 2
            else:
                total += self._sum_of_balls_in_frame(frame_index)
                frame_index += 2
        return total

    def _is_strike(self, frame_index):
        return self._rolls[frame_index] == 10

    def _is_spare(self, frame_index):
        return self._rolls[frame_index] + self._rolls[frame_index+1] == 10

    def _strike_bonus(self, frame_index):
        return self._rolls[frame_index+1] + self._rolls[frame_index+2]

    def _spare_bonus(self, frame_index):
        return self._rolls[frame_index+2]

    def _sum_of_balls_in_frame(self, frame_index):
        return self._rolls[frame_index] + self._rolls[frame_index+1]

    def _update_state(self, pins):
        strike = pins == 10
        if self._frame < 20:
            self._frame += 2 if strike else 1
        self._first = not strike if not self._first else True

        if self._frame == 20:
            if strike:
                self._extra_rolls = 2
            elif len(self._rolls) > 1 and self._rolls[-2] + self._rolls[-1] == 10:
                self._extra_rolls = 1

    def _game_over(self):
        return self._frame >= 20 and self._extra_rolls == 0